mro.1
%
%
.TH mro 1 2019-03-03 
.SH NAME
mro \- a stack-based macro processor that can form macros based on results returned by scheme functions
.SH SYNOPSIS 
.B mro

.SH DESCRIPTION 
.B mro
is a stack-based macro processor.  It reads from standard input, evaluates the macros provided to standard input, and writes the processed text to standard output.  The macros can be constructed by evaluating Guile Scheme code.  The macro language is parsed by reading one character at a time from left to right in a single pass.

.LP
The macro language is centered around a stack.  When you are on the bottom of the stack, characters typed in are echoed to standard output until a command is reached.  You can move up a level in the stack by using the PUSH command (by default: #).  Text you enter at higher levels of the stack will not be put to standard output but can be accessed by 
.B mro
commands.

.LP
Macro processing languages are used for text generation whether that text is code, a website, or writing.

.SH COMMANDS

The following are the 
.B mro
commands
.TP
.BR #
PUSH command.  All text written after the command is the next level up on the stack.
.TP
.BR @
DEFINE command.  Pops the text off the top two levels of the stack and assigns the text at the top of the text to the macro name at the second level from the top.
.TP
.BR ~
REFERENCE command.  Pops the text from the top of the stack, looks up the name of the macro specified there, and puts that text's value to the next level down in the stack.
.TP
.BR $
EXPAND command.  Expands the macros in the text on the top level of the stack.
.TP
.BR |
SHELL command.  Evalutes the text on the top level of the stack as a shell command, and outputs the output of the command to the next level down on the stack.
.TP
.BR !
CODE command.  Evaluates the text on the top level of the stack as Guile Scheme code, converts the result to a string using Guile's

.B display
function, and puts the text to the next buffer down in the stack.
.TP
.BR '='
PUSH2 command. 
.TP
.BR `text'
QUOTE text.  Put the literal text to the current level of the stack.  This is how you escape the command characters.

.SH "GUILE FUNCTIONS"
.LP
The following Guile functions are built into 
.B mro
.

.TP
.BR "(source filename)"
treats the file as if it were actually written to standard input.  In other words, 
.B mro
 parses the contents of the file.  To include the file without parsing it, use
.B "#cat filename|."
.TP
.BR "(add_to_dnp char)"
adds the character to the "do not print list". 
.B mro
will ignore this character while parsing.
.TP
.BR "(printall)"
removes all characters from the "do not print list".
.SH EXAMPLES
.LP
.BR "Section counter"
.br
The 
.B mro
commands in
.RS
#i=1@
.br
#sec=`Section #i~#i##(+ #i~ 1);@'@
.br
##sec~$
.br
##sec~$
.RE
are expanded to
.RS
Section 1
.br
Section 2
.RE
.BR "Boilerplate Guile Code in mro itself"

Guile functions in C have to be registered. 
.B mro
makes this process easy to generate.  It uses these macros in its own source.
.RS
  #register=
.br
  void*
.br
  register_guile_functions (void* data)
.br  
  {@
.br  
  #gfunc=`#register##register~
.br  
  scm_c_define_gsubr("#name~", #argnum~, 0, 0, &guile_#name~);@
.br
  SCM
.br
  guile_#name~'@
.br  
  #regbuild=`#register~
.br  
.br
  return NULL;
.br  
  }'@
.RE

Then, we can create guile functions like:
.RS
    #name=source@
.br
    #argnum=1@
.br    
    ##gfunc~$ (SCM file) { ... }
.RE

And then type
.B "##regbuild~$"
to output the
.B register_guile_functions
function that can be used with Guile's
.B scm_with_guile
function.
.SH "SEE ALSO"
.SH BUGS
.TP
Please report as an issue to https://github.com/flynnzac/mro
.SH AUTHOR
Zach Flynn <zlflynn@gmail.com>
0.0.1


mro.expand.c
/* 
   mro - a stack-based macro expander extendable with Guile Scheme

   Copyright Zach Flynn <zlflynn@gmail.com>

   This program is free software: you can redistribute it and/or modify it under
   the terms of version 3 of the GNU General Public License as published by the
   Free Software Foundation.

   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
   FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
   details.

   You should have received a copy of the GNU General Public License along with
   this program.  If not, see <https://www.gnu.org/licenses/>.
*/


/* syntax idea for "function call"
   #x#y#f:
   expands to "#1=x@#2=y@##f~$" on a separate stack.  The result is written back to the top of current stack.

   Will need to switch from one macro_stack, one buffer_stack to global and function scope.  Keep the stacks completely separate.  No fall back to global.

   "function definition" is current syntax:
   
   #f=`(+ #1~ #2~)'@
   
*/


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <libguile.h>
;

/* parser state */
static int inquote = 0;

/* buffer stack */
struct buffer { char* text; int location; int size; };
struct buffer_stack { struct buffer* buf; int level; int n_pages; };

static struct buffer_stack stack;

/* macro table */
struct macro { char* name; char* value; };
struct macro_stack { struct macro* table; int n_pages; int n_macros; };

static struct macro_stack m;

/* characters not to print */
static char* dnp;
static int n_dnp = 0;

/* do not print functions/macros */
;

int
check_dnp (int c)
{
  int i;

  for (i=1; i < n_dnp; i++)
    if (dnp[i]==c)
      return 1;

  return 0;
}

/* pops buffer off stack */
struct buffer*
pop_buffer_stack ()
{
  stack.level--;
  return &stack.buf[stack.level];
}

/* writes character to buffer */
void
push_to_buffer (struct buffer* b, int c)
{
  if (b->size <= b->location)
    {
      b->text = realloc(b->text,
                        sizeof(char)*(b->size + PAGE_BUFFER));
      b->size = b->size + PAGE_BUFFER;
    }
  
  b->text[b->location] = c;
  b->location++;
}

/* init buffer */
void
init_buffer (struct buffer* b)
{
  b->text = malloc(sizeof(char)*PAGE_BUFFER);
  b->size = PAGE_BUFFER;
  b->location = 0;
}

/* copy buffer to string */
char*
copy_from_buffer (struct buffer* src)
{
  int i;
  char* dest;
  dest = malloc(sizeof(char)*(src->location+1));
  
  for (i=0; i < src->location; i++)
    dest[i] = src->text[i];

  dest[src->location] = '\0';
  return dest;
}

/* pushes text to buffer or screen */
void
output (int c)
{
  if (check_dnp(c))
    return;
  
  if (stack.level == 0)
    printf("%c", c);
  else
    push_to_buffer(&stack.buf[stack.level-1], c);
}

/* null terminate a buffer */
void
null_terminate (struct buffer* b)
{
  push_to_buffer(b, '\0');
  b->location--;
}

/* clean up macro table */
void
free_macro_table ()
{
  int i;
  for (i=0; i < m.n_macros; i++)
    {
      free(m.table[i].name);
      free(m.table[i].value);
    }
  free(m.table);
}

/* clean up stack */
void
free_stack ()
{
  int i;
  for (i=0; i < (PAGE_STACK*stack.n_pages); i++)
    free(stack.buf[i].text);

  free(stack.buf);
}

/* look up macro name */
int
look_up_name (char* name)
{
  int i;

  for (i = 0; i < m.n_macros; i++)
    if (strcmp(m.table[i].name, name)==0)
      return i;

  return -1;
}

/* add macro to table */
void
add_macro_to_table (char* name, char* value)
{
  int loc;
  loc = look_up_name(name);
  if (loc < 0)
    {
      if (m.n_macros >= (PAGE_MACRO*m.n_pages))
	{
	  m.n_pages++;
	  m.table = realloc(m.table,
			    sizeof(struct macro)*PAGE_MACRO*m.n_pages);
	}
      m.table[m.n_macros].name = name;
      m.table[m.n_macros].value = value;
      m.n_macros++;
    }
  else
    {
      free(m.table[loc].value);
      free(name);
      m.table[loc].value = value;
    }
}


/* add macro given top two elements in buffer */
void
push_macro ()

{
  struct buffer* value;
  struct buffer* name;

   value=pop_buffer_stack(); null_terminate(value);;
   name=pop_buffer_stack(); null_terminate(name);;

  add_macro_to_table(copy_from_buffer(name),
		     copy_from_buffer(value));
  
}





/* expands macros from a file stream */
void
expand_macros (FILE* f)
{
  int i,c,loc;
  struct buffer* buf;
  char* guile_str;
  SCM guile_ret;
  FILE* f2;

  while (((c = fgetc(f)) != EOF) && c != '\0')
    {
      if (inquote)
        {
          if (c == '\'') inquote = 0; else output(c);;
	}
      else
        {
          switch (c)
            {
            case PUSH2:
              if (stack.level != 1)
                {
                  output(c);
                  break;
                }
            case PUSH:
              {
                if (stack.level >= (PAGE_STACK*stack.n_pages))
                  {
                    stack.n_pages++;
                    stack.buf = realloc(stack.buf,
                                        sizeof(struct buffer)*
                                        PAGE_STACK*stack.n_pages);
                    for (i=(stack.n_pages-1)*PAGE_STACK;
                         i < (PAGE_STACK*stack.n_pages); i++)
                      init_buffer(stack.buf+i);
                  }
                stack.buf[stack.level].location = 0;
                stack.level++;
              }
              break;
            case DEFINE:
              
              
              if (stack.level >= 2) { push_macro(); } else { output(c); } break;;
            case REF:
              
              ;
              if (stack.level >= 1) { 
              buf=pop_buffer_stack(); null_terminate(buf);;
              loc = look_up_name(buf->text);
              if (loc >= 0)
                for (i=0; i < strlen(m.table[loc].value); i++)
                  output(m.table[loc].value[i]);
               } else { output(c); } break;;
            case CODE:
              
              ;
	      
              if (stack.level >= 1) { 
              buf=pop_buffer_stack(); null_terminate(buf);;
              guile_ret = scm_c_eval_string(buf->text);
              if (!scm_is_eq(guile_ret, SCM_UNSPECIFIED))
                {
                  guile_str = scm_to_locale_string
                    (scm_object_to_string
                     (guile_ret,
                      scm_c_eval_string("display")));
                  for (i=0; i < strlen(guile_str); i++)
                    output(guile_str[i]);
            
                  free(guile_str);
                } } else { output(c); } break;;
            case EXPAND:
              
              ;
              if (stack.level >= 1) { 
              buf=pop_buffer_stack(); null_terminate(buf);;
              f2 = fmemopen(buf->text, buf->size, "r");
              expand_macros(f2);
              fclose(f2); } else { output(c); } break;;
            case SHELL:
              
              ;
              if (stack.level >= 1) { 
              buf=pop_buffer_stack(); null_terminate(buf);;
              f2 = popen(buf->text, "r");
              while (((c=fgetc(f2)) != EOF) && c != '\0')
		output(c);
              pclose(f2); } else { output(c); } break;;
            case '`':
              inquote = 1;
              break;
            default:
              output(c);
              break;
            }
        }
    }
}

/* define macros to add guile functions */

    
      ;

/* guile: add to do not print list */




    SCM
      guile_add_to_dnp (SCM ch)
{
  char* str = scm_to_locale_string(ch);
   
dnp = realloc(dnp, sizeof(char)*(n_dnp+1));
dnp[n_dnp] = str[0];
n_dnp++;

  free(str);
  return scm_from_locale_string("");
}

/* guile: clear do not print list */




    SCM
      guile_printall ()
{
  dnp = realloc(dnp, sizeof(char));
  dnp[0] = '\0';
  n_dnp = 1;

  return scm_from_locale_string("");
}

/* guile: start a definition section by adding \n to do not print
   list */




    SCM
      guile_defsec ()
{
   
dnp = realloc(dnp, sizeof(char)*(n_dnp+1));
dnp[n_dnp] = '\n';
n_dnp++;

  return scm_from_locale_string("");
}

/* guile: source a macro file as if it was entered along with text */



    SCM
      guile_source (SCM file)
{
  char* file_c = scm_to_locale_string(file);
  FILE* f = fopen(file_c, "r");
  expand_macros(f);
  fclose(f);
  free(file_c);
  return scm_from_locale_string("");
}


void*
register_guile_functions (void* data)
{
    scm_c_define_gsubr("add_to_dnp", 1, 0, 0, &guile_add_to_dnp);
    scm_c_define_gsubr("printall", 0, 0, 0, &guile_printall);
    scm_c_define_gsubr("defsec", 0, 0, 0, &guile_defsec);
    scm_c_define_gsubr("source", 1, 0, 0, &guile_source);

      return NULL;
}

/* main program */
int
main (int argc, char** argv)
{
  int i;
  char* val;
  char* num;
  
  stack.level = 0;
  stack.n_pages = 1;
  stack.buf = malloc(sizeof(struct buffer)*PAGE_STACK);
  for (i=0; i < PAGE_STACK; i++)
    init_buffer(stack.buf+i);

  m.n_macros = 0;
  m.n_pages = 1;
  m.table = malloc(sizeof(struct macro)*PAGE_MACRO);

  dnp = malloc(sizeof(char));
  dnp[0] = '\0';
  n_dnp = 1;
  scm_with_guile(&register_guile_functions, NULL);

  /* expand , , ... to the command line args */
  for (i=1; i < argc; i++)
    {
      num = malloc(sizeof(char)*((i / 10) + 1+1));
      sprintf(num, "%d", i-1);

      val = malloc(sizeof(char)*(strlen(argv[i])+1));
      strcpy(val, argv[i]);
      
      add_macro_to_table(num,val);
    }

  expand_macros(stdin);

  free_stack();
  free_macro_table();
  free(dnp);

  return 0;
}
